scripts/model.py

```diff
@@ -0,0 +1,127 @@
+import torch
+from transformers import AutoTokenizer, GPTJForCausalLM
+import torch.nn.functional as F
+
+from utils import cached_decode_tok
+
+
+class ModelWatcher:
+    def __init__(self):
+
+        num_layers = 28
+        device_ids = list(range(torch.cuda.device_count()))
+        num_layers_per_device = num_layers // len(device_ids)
+        device_map = { i: list(range(i * num_layers_per_device : (i + 1) * num_layers_per_device)) for i in device_ids }
+        self.devices = [torch.device(f'cuda:{i}') for i in device_ids]
+
+        print('Loading model...')
+        self.tokenizer = AutoTokenizer.from_pretrained("EleutherAI/gpt-j-6B")
+        self.model = GPTJForCausalLM.from_pretrained("EleutherAI/gpt-j-6B", low_cpu_mem_usage=True)
+        print(f'Parallelizing on {len(device_ids)} GPUs...')
+        self.model.parallelize(device_map)
+        self.model.eval()
+
+        print('Setting up handlers...')
+        self.module_names_to_track_for_activations = \
+            [f'transformer.h.{i}.mlp.fc_out' for i in range(self.model.config.n_layer)]
+        self.module_names_to_track_for_logits = \
+            [f'transformer.h.{i}' for i in range(self.model.config.n_layer)]
+
+        self.hidden_states = dict()
+
+        self.handles = []
+        cnt = 0
+        for name, m in self.model.named_modules():
+            if name in self.module_names_to_track_for_activations or name in self.module_names_to_track_for_logits:
+                cnt += 1
+                handle = m.register_forward_hook(partial(self._save_hidden_states, name))
+                self.handles.append(handle)
+
+    def _save_hidden_states(self, name, module, input, output):
+        # we care about input for activations, and output for logit lens
+        hidden_states[name] = { 'input': input[0], 'output': output }
+
+    def clear_hidden_states(self):
+        self.hidden_states = dict()
+
+    def extract_neuron_values(self, threshold):
+        """
+        For each MLP, determine which neurons fire at any point during the entire sequence,
+        unless it only fires on the first token (which we will just assume is noise).
```

Comment:
I think that I removed the "ignore first neuron" filter based on what we discussed on Discord, so that's no longer consistent with this comment.